### PR TITLE
process monitor telemetry

### PR DESCRIPTION
**Pull request by nplanel**
### What does this PR do?

Add internal telemetry for process monitor

process monitor will process :
 o events refer to netlink events received (not only exec and exit)
 o exec process netlink events
 o exit process netlink events
 o restart the netlink connection

 o reinit_failed is the number of failed re-initialisation after a netlink restart due to an error
 o process_scan_failed would be > 0 if initial process scan failed
 o callback_executed numbers of executed callback 

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/process/monitor/process_monitor.go [hunk 1]
```diff
@@ -35,6 +36,47 @@ var (
...
+	pmt.events = metricGroup.NewMetric("events")
+	pmt.exec = metricGroup.NewMetric("exec")
+	pmt.exit = metricGroup.NewMetric("exit")
+	pmt.restart = metricGroup.NewMetric("restart")
+
+	pmt.refcount = metricGroup.NewMetric("refcount")
+	pmt.refcount.Add(refcount)
+	pmt.reinitFailed = metricGroup.NewMetric("reinitFailed")
+	pmt.processScanFailed = metricGroup.NewMetric("processScanFailed")
+	pmt.callbackCalled = metricGroup.NewMetric("callbackCalled")
```
Comment (guyarb) [inline pkg/process/monitor/process_monitor.go:77]: use lower_camel_case_with_underscores for the metric names

File: pkg/process/monitor/process_monitor.go [hunk 2]
```diff
@@ -35,6 +36,47 @@ var (
...
+	//  o refcount is number of used monitor by different subsystem
+	//  o reinitFailed is the number of failed re-initialisation after a netlink restart due to an error
+	//  o processScanFailed would be > 0 if initial process scan failed
+	//  o callbackCalled numbers of callback called
+	events  *telemetry.Metric
+	exec    *telemetry.Metric
+	exit    *telemetry.Metric
+	restart *telemetry.Metric
+
+	refcount          *telemetry.Metric
```
Comment (guyarb) [inline pkg/process/monitor/process_monitor.go:56]: wondering why do we need that metric
the number is eventually equals to the number of modules use process monitor (won't during after initialization of the agent) and we know what components are enabled

Comment (guyarb) [reply pkg/process/monitor/process_monitor.go:56]: I think we should remove it

File: pkg/process/monitor/process_monitor.go [hunk 3]
```diff
@@ -186,10 +234,12 @@ func (pm *ProcessMonitor) mainEventLoop() {
 			}
 
 			pm.eventCount.Inc()
+			pm.tel.events.Add(1)
```
Comment (guyarb) [inline pkg/process/monitor/process_monitor.go:237]: why do we need both?

File: pkg/process/monitor/process_monitor.go [hunk 4]
```diff
@@ -186,10 +234,12 @@ func (pm *ProcessMonitor) mainEventLoop() {
...
 
 			switch ev := event.Msg.(type) {
 			case *netlink.ExecProcEvent:
 				pm.execCount.Inc()
+				pm.tel.exec.Add(1)
```
Comment (guyarb) [inline pkg/process/monitor/process_monitor.go:242]: why do we need both?

File: pkg/process/monitor/process_monitor.go [hunk 5]
```diff
@@ -198,6 +248,7 @@ func (pm *ProcessMonitor) mainEventLoop() {
 				}
 			case *netlink.ExitProcEvent:
 				pm.exitCount.Inc()
+				pm.tel.exit.Add(1)
```
Comment (guyarb) [inline pkg/process/monitor/process_monitor.go:251]: why do we need both?

File: pkg/process/monitor/process_monitor.go [hunk 6]
```diff
@@ -210,6 +261,7 @@ func (pm *ProcessMonitor) mainEventLoop() {
 				return
 			}
 			pm.restartCounter.Inc()
+			pm.tel.restart.Add(1)
```
Comment (guyarb) [inline pkg/process/monitor/process_monitor.go:264]: why do we need both?

**Review by guyarb — CHANGES_REQUESTED** [review]
CRed

File: pkg/process/monitor/process_monitor.go (hunk 2 above)
Comment (nplanel) [reply pkg/process/monitor/process_monitor.go:56]: yeah I removed it as should not hit something bad here

File: pkg/process/monitor/process_monitor.go (hunk 3 above)
Comment (nplanel) [reply pkg/process/monitor/process_monitor.go]: eventCount got logged and reseted every 2 minutes, I merged and use the telemetery logic to log in the console

File: pkg/process/monitor/process_monitor.go [hunk 7]
```diff
@@ -165,7 +202,6 @@ func (pm *ProcessMonitor) mainEventLoop() {
 	logTicker := time.NewTicker(2 * time.Minute)
 
 	defer func() {
-		logTicker.Stop()
```
Comment (guyarb) [inline pkg/process/monitor/process_monitor.go:168]: why did you remove the stopping of the log ticket?

**Review by guyarb — CHANGES_REQUESTED** [review]
CRed

File: pkg/process/monitor/process_monitor.go [hunk 8]
```diff
@@ -35,6 +36,45 @@ var (
...
+	)
+	pmt.events = metricGroup.NewMetric("events", telemetry.OptMonotonic)
+	pmt.exec = metricGroup.NewMetric("exec", telemetry.OptMonotonic)
+	pmt.exit = metricGroup.NewMetric("exit", telemetry.OptMonotonic)
+	pmt.restart = metricGroup.NewMetric("restart", telemetry.OptMonotonic)
+
+	pmt.eventsChannelSize = metricGroup.NewMetric("events_channel_size", telemetry.OptGauge)
+	pmt.reinitFailed = metricGroup.NewMetric("reinit_failed", telemetry.OptMonotonic)
+	pmt.processScanFailed = metricGroup.NewMetric("process_scan_failed", telemetry.OptMonotonic)
+	pmt.callbackCalled = metricGroup.NewMetric("callback_called", telemetry.OptMonotonic)
```
Comment (guyarb) [inline pkg/process/monitor/process_monitor.go:75]: suggesting to rename into "callback_executed"

File: pkg/process/monitor/process_monitor.go [hunk 9]
```diff
@@ -185,19 +222,20 @@ func (pm *ProcessMonitor) mainEventLoop() {
 				return
 			}
 
-			pm.eventCount.Inc()
+			pm.tel.events.Add(1)
+			pm.tel.eventsChannelSize.Set(int64(len(pm.netlinkEventsChannel)))
```
Comment (guyarb) [inline pkg/process/monitor/process_monitor.go:226]: can you show how the metric looks like?
I'm not sure if the measuring of the events channel makes sense

Comment (guyarb) [reply pkg/process/monitor/process_monitor.go:226]: it might be more relevant to set it every interval (like inside the logTicker case)

**Review by guyarb — DISMISSED** [review]
CRed

File: pkg/process/monitor/process_monitor.go (hunk 9 above)
Comment (nplanel) [reply pkg/process/monitor/process_monitor.go:226]: the previous chan log was useless as it's mostly reported 0 or if it was a positive value it would be a very loaded environment.
I change the local counter to max chan size and report it every 2 minutes (ticker) only on log

File: pkg/process/monitor/process_monitor.go [hunk 10]
```diff
@@ -35,6 +36,43 @@ var (
...
+
+	reinitFailed      *telemetry.Counter
+	processScanFailed *telemetry.Counter
+	callbackExecuted  *telemetry.Counter
+}
+
+func (pmt *processMonitorTelemertry) initialize() {
+	metricGroup := telemetry.NewMetricGroup(
+		"process.monitor",
+		telemetry.OptPayloadTelemetry,
```
Comment (guyarb) [inline pkg/process/monitor/process_monitor.go:64]: not all telemetry should be with `OptPayloadTelemetry`
@p-lambert please review and give your suggestions

File: pkg/process/monitor/process_monitor.go [hunk 11]
```diff
@@ -35,6 +36,43 @@ var (
...
+	restart *telemetry.Counter
+
+	reinitFailed      *telemetry.Counter
+	processScanFailed *telemetry.Counter
+	callbackExecuted  *telemetry.Counter
+}
+
+func (pmt *processMonitorTelemertry) initialize() {
+	metricGroup := telemetry.NewMetricGroup(
+		"process.monitor",
```
Comment (guyarb) [inline pkg/process/monitor/process_monitor.go:63]:
```suggestion
		"usm.process.monitor",
```

Comment (guyarb) [reply pkg/process/monitor/process_monitor.go:63]: the module is used only by USM

File: pkg/process/monitor/process_monitor.go [hunk 12]
```diff
@@ -185,19 +221,22 @@ func (pm *ProcessMonitor) mainEventLoop() {
 				return
 			}
 
-			pm.eventCount.Inc()
+			if maxChannelSize < len(pm.netlinkEventsChannel) {
+				maxChannelSize = len(pm.netlinkEventsChannel)
+			}
```
Comment (guyarb) [inline pkg/process/monitor/process_monitor.go:226]: not sure if that gives us any valuable information.
furthermore I have a concern about the performance impact of that

File: pkg/process/monitor/process_monitor.go (hunk 9 above)
Comment (guyarb) [reply pkg/process/monitor/process_monitor.go]: moving the discussion to https://github.com/DataDog/datadog-agent/pull/18018#discussion_r1274704913

File: pkg/process/monitor/process_monitor.go (hunk 12 above)
Comment (nplanel) [reply pkg/process/monitor/process_monitor.go:226]: internally chan size is maintained `qcount` https://github.com/golang/go/blob/master/src/runtime/chan.go#L34C1
and len(chan) will call https://github.com/golang/go/blob/master/src/runtime/chan.go#L728

File: pkg/process/monitor/process_monitor.go (hunk 10 above)
Comment (nplanel) [reply pkg/process/monitor/process_monitor.go]: I ask @p-lambert and changed to Prometheus

File: pkg/process/monitor/process_monitor.go [hunk 13]
```diff
@@ -219,11 +257,18 @@ func (pm *ProcessMonitor) mainEventLoop() {
...
 			}
 		case <-logTicker.C:
-			log.Debugf("process monitor stats - total events: %d; exec events: %d; exit events: %d; Channel size: %d; restart counter: %d",
-				pm.eventCount.Swap(0), pm.execCount.Swap(0), pm.exitCount.Swap(0), len(pm.netlinkEventsChannel), pm.restartCounter.Load())
+			log.Debugf("process monitor stats - total events: %d; exec events: %d; exit events: %d; restart counter: %d; max channel size: %d / 2 minutes)",
+				pm.tel.events.Get(),
+				pm.tel.exec.Get(),
+				pm.tel.exit.Get(),
+				pm.tel.restart.Get(),
+				maxChannelSize,
```
Comment (p-lambert) [inline pkg/process/monitor/process_monitor.go:269]: you can get this for free by using `MetricGroup.Summary()`

File: pkg/process/monitor/process_monitor_test.go [hunk 14]
```diff
@@ -179,4 +214,20 @@ func TestProcessMonitorInNamespace(t *testing.T) {
 		_, captured := execSet.Load(cmd.ProcessState.Pid())
 		return captured
 	}, time.Second, 200*time.Millisecond, "did not capture process EXEC from other namespace")
+
+	tel := telemetry.ReportPayloadTelemetry("1")
```
Comment (p-lambert) [inline pkg/process/monitor/process_monitor_test.go:218]: I don't think this is the proper way to test this.
This function will only return things instantiated with `OptPayloadTelemetry`,  and your metrics are all `OptPrometheus`.

In order to decouple your tests from the metric "export" type, I'd simply write assertions like 
```
assert.Equal(t, 1, pm.tel.callbackExecuted.Get())
```
Just make sure to call `telemetry.Clear()` at the beginning of tests because these counters are monotonic

File: pkg/process/monitor/process_monitor.go [hunk 15]
```diff
@@ -35,6 +36,42 @@ var (
...
+	exec    *telemetry.Counter
+	exit    *telemetry.Counter
+	restart *telemetry.Counter
+
+	reinitFailed      *telemetry.Counter
+	processScanFailed *telemetry.Counter
+	callbackExecuted  *telemetry.Counter
+}
+
+func (pmt *processMonitorTelemertry) initialize() {
```
Comment (p-lambert) [inline pkg/process/monitor/process_monitor.go:60]: I feel this is an anti-pattern in Go. why not having a constructor function that returns a `processMonitorTelemetry ` instance?

File: pkg/process/monitor/process_monitor.go [hunk 16]
```diff
@@ -35,6 +36,42 @@ var (
 	}
 )
 
+// processMonitorTelemertry
+type processMonitorTelemertry struct {
```
Comment (p-lambert) [inline pkg/process/monitor/process_monitor.go:40]: This typo is being propagated in the whole change 😅

**Review by p-lambert — CHANGES_REQUESTED** [review]

File: pkg/process/monitor/process_monitor.go (hunk 15 above)
Comment (nplanel) [reply pkg/process/monitor/process_monitor.go:60]: I moved to newProcessMonitorTelemetry()

File: pkg/process/monitor/process_monitor.go (hunk 13 above)
Comment (nplanel) [reply pkg/process/monitor/process_monitor.go:269]: Thanks

File: pkg/process/monitor/process_monitor_test.go (hunk 14 above)
Comment (nplanel) [reply pkg/process/monitor/process_monitor_test.go]: I added a return to ReportPrometheus()
This way I would check end to end the value are correct with the correct exported name

File: pkg/process/monitor/process_monitor_test.go [hunk 17]
```diff
@@ -75,6 +80,35 @@ func TestProcessMonitorSanity(t *testing.T) {
...
+			return 0
+		}
+		switch v := pm.(type) {
+		case libtelemetry.Counter:
+			return v.WithValues().Get()
+		case libtelemetry.Gauge:
+			return v.WithValues().Get()
+		}
+		return 0
+	}
```
Comment (p-lambert) [inline pkg/process/monitor/process_monitor_test.go:96]: Can you please re-write the tests following my previous suggestion of relying on the public API of the library?

Example:
```
assert.Equal(t, pm.telemetry.metricName.Get(), 1)
```

Not only the tests are unnecessarily complex at the moment, but they are completely coupled to Prometheus, and one of the goals of the `telemetry` library is precisely abstracting away the details of how data gets reported. In other words, changing a metric from `Prometheus` to `PayloadTelemtry` should *not* break your tests.

**Review by p-lambert — CHANGES_REQUESTED** [review]

File: pkg/process/monitor/process_monitor_test.go (hunk 17 above)
Comment (nplanel) [reply pkg/process/monitor/process_monitor_test.go:96]: As discuss offline let's not test metrics end to end.
We could add a test in the telemetry test suite

**Review by p-lambert — APPROVED** [review]
LGTM once CI is green.
Thanks for updating the PR!

**Comment by nplanel** [comment]
@guyarb len(chan) is still an issue on your side ?

**Comment by guyarb** [comment]
> @guyarb len(chan) is still an issue on your side ?

no
